Support DoubleValues expression scripts in lang-expression

grcevski opened the pull request:
This PR adds support for getting custom DoubleValues scripts and an implementation with lucene expressions. With this change plugins can implement any lucene DoubleValues custom logic.

Commit:
Initial code to support binary expr scripts

elasticsearchmachine commented:
Hi @grcevski, I've created a changelog YAML for you.

grcevski committed:
Update docs/changelog/89895.yaml

jdconrad commented on server/src/main/java/org/elasticsearch/script/BinaryScript.java:
```diff
@@ -0,0 +1,34 @@
+/**
+ * A string template rendered as a script.
+ */
+public abstract class BinaryScript<T> {
```
I wonder if RawScript makes more sense than BinaryScript?

jdconrad commented on server/src/main/java/org/elasticsearch/script/BinaryScript.java:
```diff
@@ -0,0 +1,34 @@
+
+import java.io.IOException;
+
+/**
+ * A string template rendered as a script.
+ */
```
Not sure this is considered a template?

grcevski replied:
Ugh, copy pasta :/

jdconrad commented on server/src/main/java/org/elasticsearch/script/BinaryScript.java:
```diff
@@ -0,0 +1,34 @@
+    public abstract T execute();
+
+    /** A factory to construct {@link BinaryScript} instances. */
+    public interface LeafFactory<T> {
+        BinaryScript<T> newInstance() throws IOException;
+    }
+
+    public interface Factory<T> extends ScriptFactory {
+        BinaryScript.LeafFactory<T> newFactory();
+    }
```
Since this script won't be using leaf segments directly (right?) we can probably skip the leaf factory, so there's only a single level of indirection.

grcevski replied:
Sounds good.

jdconrad commented on server/src/main/java/org/elasticsearch/script/BinaryScript.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    @SuppressWarnings("rawtypes")
+    public static final ScriptContext<Factory> CONTEXT = new ScriptContext<>("binary", Factory.class);
```
Wonder if maybe "raw" makes more sense than "binary"?

jdconrad commented on modules/lang-expression/src/main/java/org/elasticsearch/script/expression/ExpressionBinaryScript.java:
```diff
@@ -0,0 +1,37 @@
+
+import java.io.IOException;
+
+public class ExpressionBinaryScript implements BinaryScript.LeafFactory<Expression> {
```
Maybe RawExpressionScript?

grcevski replied:
I like the Raw naming.

jdconrad reviewed (approved):
Left a few comments. Thanks for taking this on!

Commit:
Address PR feedback

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

Commit:
Merge branch 'main' into plugin/custom_expression

Commit:
Wrap the raw expression

This allows the raw expression to be used while
crossing class loader boundaries, since lang-expressions
from Lucene is not part of server.

Commit:
spotless

Commit:
Merge branch 'main' into plugin/custom_expression

Commit:
fix missing method

Commit:
Clean-up comments and naming

Added few tests for the new functions.

Commit:
Add ways for tests to create classloaders

ChrisHegarty reviewed (approved):
LGTM.

jdconrad reviewed (approved):
This is great! Thanks for putting this together @grcevski
